packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/exceptions/AppExceptions.kt

```diff
@@ -0,0 +1,237 @@
+ *   - [SyncException]
+ *     - [UnrecoverableSyncException]
+ *     - [WrongSyncTypeException]
+ *
+ * This hierarchy is intended to model errors in a way so exceptions at the bottom of the hierarchy
+ * are _actionable_, i.e. it should be clear from the exception which action can be taken to
+ * resolve it.
+ *
+ * Exceptions further up the hierarchy are used as a way to categorize the errors and are
+ * harder to react to in a single way that will make sense to end users of an app, but they should
+ * be logged for later inspection.
+ *
+ * In most cases, only exceptions at the bottom of the hierarchy will be documented in the API
+ * documentation, with one notable special case: [ConnectionException]. This exception is
+ * assumed to be thrown by all methods that mentions a [ServiceException] and covers all transport
+ * errors. These are often intermittent, so catching this exception and retrying should generally
+ * be safe, but more information can be found in the documentation for [ConnectionException].
+ **
+ * With the above exception hierarchy in mind, a sensible way to handle errors could look like
+ * this:
+ *
+ * ```
+ * val app = App.with("my-app-id")
+ * runCatching {
+ *     app.login(Credentials.emailPassword("myemail@mail.com", "mypassword"))
+ * }.onSuccess {
+ *     gotoMainScreen()
+ * }.onFailure { ex: Throwable ->
+ *     when(ex) {
+ *         is InvalidCredentialsException -> {
+ *             showWrongPasswordDialog()
+ *         }
+ *         is ConnectionException -> {
+ *             CrashLogger.log(ex.toString())
+ *             showRetryLoginDialog()
+ *         }
+ *         else -> {
+ *             CrashLogger.log(ex.toString())
+ *             showContactAdministratorDialog()
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * For more information about specific exceptions, see the documentation for that exception.
+ *
+ * @see ServiceException
+ * @see SyncException
+ */
+public open class AppException : RealmException {
+    internal constructor(message: String) : super(message)
+}
+
+/**
+ * This exception is considered the top-level or "catch-all" for problems related to HTTP requests
+ * made towards Atlas App Services. This covers both HTTP transport problems, problems passing JSON
+ * or the server considering the request invalid, for whatever reason.
+ *
+ * Generally, reacting to this exception will be hard, except to log the error for further
+ * analysis. But in many cases a more specific subtype will be thrown, which will be easier to react
+ * to. See the subclasses of this exception for more information about these.
+ *
+ * @see ConnectionException
+ * @see BadRequestException
+ * @see AuthException
+ */
+public open class ServiceException : AppException {
+    internal constructor(message: String) : super(message)
+}
+
+/**
+ * Exception indicating that something went wrong with the underlying HTTP request towards one of
+ * the Atlas App Services. The exact cause is in the exception message.
```

Comment:
```suggestion
 * Exception indicating that something went wrong with the underlying HTTP request to 
 * Atlas App Services. The exact cause is in the exception message.
```
We're still settling our language around "Atlas App Services" for the documentation site, but I suspect we won't refer to "Services" as individual items much if ever (since they're all part of one product).  This should be a little more consistent with the language we'll use when we complete that refactor.